### PR TITLE
fixes #26250 - add dns_servers helper

### PR DESCRIPTION
**Pull request by timogoebel**

### CODE REVIEW

File: test/models/subnet_test.rb [hunk 1]
```diff
@@ -144,4 +144,21 @@ class Subnet::Test < Subnet; end
   test "should have MTU set to 1500 by default" do
     assert_equal 1500, Subnet.new.mtu
   end
+
+  describe '#dns_servers' do
+    test 'should display a list of dns servers' do
+      subnet = FactoryBot.create(:subnet_ipv4, dns_primary: '8.8.8.8', dns_secondary: '8.8.4.4')
```
Comment (ekohl) [inline test/models/subnet_test.rb]: I always prefer to use IPs reserved for documentation, examples and tests. https://tools.ietf.org/html/rfc5735#section-3 states:
> 192.0.2.0/24 - This block is assigned as "TEST-NET-1" for use in documentation and example code. It is often used in conjunction with domain names example.com or example.net in vendor and protocol documentation. As described in [RFC5737], addresses within this block do not legitimately appear on the public Internet and can be used without any coordination with IANA or an Internet registry. See [RFC1166].

There's also TEST-NET-2 (198.51.100.0/24) and TEST-NET-3 (203.0.113.0/24).

Comment (timogoebel) [reply test/models/subnet_test.rb]: As always, you are right. Let me change this.

**Review by ekohl — APPROVED** [review]

**Review by ares — CHANGES_REQUESTED** [review]
I'm sorry to be that annoying guy, but would it be possible to make the macro a bit more sophisticated?

something like we have for [host attributes](https://github.com/theforeman/foreman/blob/develop/lib/foreman/renderer/scope/macros/host_template.rb#L20-L23)

```
subnet_dns_servers(subnet)
```

that would check the subnet exists and if not raises nicer error message. IIUC, the primary usage of this is `@host.subnet.dns_servers` which will result in undefined method dns_servers for nil. In ideal world I would expect having two macros added, the one that I mentioned above and then `host_dns_servers(host)` that would use `subnet_dns_servers(host.subnet)` inside.

This would be better also for case when we decide to prefer subnet6 instead of subnet, the macro would change just internally.

**Comment by timogoebel** [comment]
> I'm sorry to be that annoying guy, but would it be possible to make the macro a bit more sophisticated?

I think that's totally out of the scope of this PR. I personally dislike all the magic template macros as explicit code is much easier to understand.

This is magic where we need to document the behavior. I'm actually not sure how I would implement this.

```ruby
host_dns_servers(host).join(', ')
```

This is pure ruby and I believe easier to understand for engineers.

```ruby
host.subnet.dns_servers.join(', ')
```

**Comment by ekohl** [comment]
I think @ares was suggesting to use it:
```ruby
subnet_dns_servers(@host.subnet)
subnet_dns_servers(@host.subnet6)
```

That would handle nil values for the subnet as well and return an empty list.

Currently I'm going over the provisioning templates and it looks like the current implementation is good enough. PR incoming.

**Comment by ekohl** [comment]
See https://github.com/theforeman/community-templates/pull/562

**Comment by timogoebel** [comment]
@ares and I talked about this on IRC and he suggested to add a macro `keep_present_values` that would just do `object.select(&:present?)`. I like the idea and will open a PR for that, but I still believe we should merge this PR. We can continue in any direction from there.

**Comment by ekohl** [comment]
I agree that being able to treat DNS servers as a list and forgetting `dns_primary` and `dns_secondary` exist is a great benefit in itself. This is how most tools think about it.

**Review by ares — COMMENTED** [review]
Yeah, though I disagree with using our internal objects API as also template DSL, I don't want to block this. Reads well, I'll give it a try and merge if no one beats me to it.

**Review by ares — APPROVED** [review]
Thanks @timogoebel, works fine. Merging! Would you mind adding it to https://projects.theforeman.org/projects/foreman/wiki/TemplateWriting ?

**Comment by timogoebel** [comment]
> Would you mind adding it to https://projects.theforeman.org/projects/foreman/wiki/TemplateWriting ?

Thanks, @ares for mering despite your concern. Added the method to the wiki documentation.

**Pull request merged**